lib/zanzibar.rb

```diff
@@ -1,210 +1,200 @@
+    # @param [Integer] the secret id
+    # @return [Hash] the secret hash retrieved from the wsdl
+
+    def get_secret(scrt_id, token = nil)
+      begin
+        secret = @@client.call(:get_secret, message: { token: token || get_token, secretId: scrt_id}).hash[:envelope][:body][:get_secret_response][:get_secret_result]
+        raise "There was an error getting secret #{scrt_id}: #{secret[:errors][:string]}" if secret[:errors]
+        return secret
+      rescue Savon::Error => err
+        raise "There was an error getting the secret with id #{scrt_id}: #{err}"
+      end
+    end
+
+    ## Retrieve a simple password from a secret
+    # Will raise an error if there are any issues
+    # @param [Integer] the secret id
+    # @return [String] the password for the given secret
+
+    def get_password(scrt_id)
+      begin
+        secret = get_secret(scrt_id)
+        secret_items = secret[:secret][:items][:secret_item]
+        return get_secret_item_by_field_name(secret_items,"Password")[:value]
+      rescue Savon::Error => err
+        raise "There was an error getting the password for secret #{scrt_id}: #{err}"
+      end
+    end
+
+    def write_secret_to_file(path, secret_response)
+      File.open(File.join(path, secret_response[:file_name]), 'wb') do |file|
+        file.puts Base64.decode64(secret_response[:file_attachment])
+      end
+    end
+
+    def get_secret_item_by_field_name(secret_items, field_name)
+      secret_items.each do |item|
+        return item if item[:field_name] == field_name
+      end
+    end
+
+    ## Get the secret item id that relates to a key file or attachment.
+    # Will raise on error
+    # @param [Integer] the secret id
+    # @param [String] the type of secret item to get, one of privatekey, publickey, attachment
+    # @return [Integer] the secret item id
+
+    def get_scrt_item_id(scrt_id, type, token)
+      secret = get_secret(scrt_id, token)
+      secret_items = secret[:secret][:items][:secret_item]
+      begin
+        return get_secret_item_by_field_name(secret_items, type)[:id]
+      rescue
+        raise "Unknown type, #{type}."
+      end
+    end
+
+    ## Downloads the private key for a secret and places it where Zanzibar is running, or :path if specified
+    # Raise on error
+    # @param [Hash] args, :scrt_id, :scrt_item_id - optional, :path - optional
+
+    def download_private_key(args = {})
+      token = get_token
+      FileUtils.mkdir_p(args[:path]) if args[:path]
+      path = args[:path] ? args[:path] : '.' ## The File.join below doesn't handle nils well, so let's take that possibility away.
+      begin
+        response = @@client.call(:download_file_attachment_by_item_id, message: { token: token, secretId: args[:scrt_id], secretItemId: args[:scrt_item_id] || get_scrt_item_id(args[:scrt_id], 'Private Key', token)}).hash[:envelope][:body][:download_file_attachment_by_item_id_response][:download_file_attachment_by_item_id_result]

```

Comment:
Looks like a good bit of duplication between the `download_*` functions (or is the hash-walking different? there's so much I can't tell). Maybe these should just be partial invocations of a single `download_attachment` function with an `args[:attachment_name]` provided for you?